consensus/src/dag/dag.rs

```diff
@@ -0,0 +1,268 @@
+            .unwrap_or(&0)
+    }
+
+    fn highest_round(&self) -> Round {
+        *self
+            .nodes_by_round
+            .iter()
+            .next_back()
+            .map(|(round, _)| round)
+            .unwrap_or(&0)
+    }
+
+    pub fn add_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+        let node = Arc::new(node);
+        let index = *self
+            .author_to_index
+            .get(&node.metadata.author)
+            .ok_or_else(|| anyhow!("unknown author"))?;
+        let round = node.metadata.round;
+        ensure!(round >= self.lowest_round(), "round too low");
+        ensure!(round <= self.highest_round() + 1, "round too high");
+        for parent in &node.parents {
+            ensure!(self.exists(&parent.digest), "parent not exist");
+        }
+        ensure!(
+            self.nodes_by_digest
+                .insert(node.metadata.digest, node.clone())
+                .is_none(),
+            "duplicate node"
+        );
+        assert!(
+            self.nodes_by_round
+                .entry(round)
+                .or_insert(vec![None; self.author_to_index.len()])[index]
+                .replace(node.clone())
+                .is_none(),
+            "equivocate node"
+        );
+        if round
+            > self.highest_unlinked_nodes_by_author[index]
+                .as_ref()
+                .map_or(0, |node| node.metadata.round)
+        {
+            self.highest_unlinked_nodes_by_author[index].replace(node);
+        }
+        Ok(())
+    }
+
+    pub fn exists(&self, digest: &HashValue) -> bool {
+        self.nodes_by_digest.get(digest).is_some()
+    }
+
+    pub fn get_node(&self, digest: &HashValue) -> Option<Arc<CertifiedNode>> {
+        self.nodes_by_digest.get(digest).cloned()
+    }
+
+    pub fn get_unlinked_nodes_for_new_round(
+        &self,
+        validator_verifier: &ValidatorVerifier,
+    ) -> Option<Vec<NodeMetadata>> {
+        let current_round = self.highest_round();
+        let strong_link_authors =
+            self.highest_unlinked_nodes_by_author
+                .iter()
+                .filter_map(|maybe_node| {
+                    maybe_node.as_ref().map_or(None, |node| {
+                        if node.metadata.round == current_round {
+                            Some(&node.metadata.author)
+                        } else {
+                            None
+                        }
+                    })
+                });
+        if validator_verifier
+            .check_voting_power(strong_link_authors)
+            .is_ok()
+        {
+            Some(
+                self.highest_unlinked_nodes_by_author
+                    .iter()
+                    .filter_map(|maybe_node| maybe_node.as_ref().map(|node| node.metadata.clone()))
+                    .collect(),
+            )
+        } else {
+            None
+        }
+    }
+
+    pub fn mark_nodes_linked(&mut self, node_metadata: &Vec<NodeMetadata>) {
```

Comment:
Q: when is this fn useful?

Reply:
I was thinking we mark it after reliable broadcast succeeds